### PR TITLE
[ROCm] ROCm-specific GPU utility functions.

### PR DESCRIPTION
**Pull request by whchung**
Common utility helpers between CUDa and ROCm are renamed:
- core/util/cuda_device_functions.h -> core/util/gpu_device_functions.h
- core/util/cuda_kernel_helper.h -> core/util/gpu_kernel_helper.h
- core/util/cuda_kernel_helper_test.cu.cc -> core/util/gpu_kernel_helper_test.cu.cc
- core/util/cuda_launch_config.h -> core/util/gpu_launch_cnofig.h

ROCm-specific changes for GPU device functions are also introduced in this PR.

GPU kernel invocation logic would be introduced in the next PR. All custom CUDA kernels are still launched as `CudaLaunchKernel` for now.

### CODE REVIEW

**Review by smit-hinsu — DISMISSED** [review]
If we go with the rename then the tensorrt change looks good to me.

**Comment by whchung** [comment]
@smit-hinsu Thanks for approving this PR. Failed CI targets are:
- MacOS Contrib
- MacOS Python2 and CC
- Windows Bazel
- Windows Bazel GPU

From the logs they seem to have nothing to do with this PR, but a configuration issue in CI test infrastructure. I'm wondering if it's possible to manually apply `ready-to-pull` label?

**Comment by ymodak** [comment]
@whchung Can you please resolve the conflict? Thanks!

**Comment by ymodak** [comment]
@whchung Can you please resolve the conflict? Thanks!

**Comment by rthadur** [comment]
> @whchung Can you please resolve the conflict? Thanks!

Gentle ping

**Comment by whchung** [comment]
@rthadur could you give another go for this PR? thanks. the failures on "Ubuntu Python3 PIP" and "Windows Bazel" don't seem to be related to this PR.

**Comment by whchung** [comment]
ping?

**Comment by rthadur** [comment]
@whchung request to re base your branch

**Comment by whchung** [comment]
@rthadur sorry for being late. I've rebased the PR.

**Review by smit-hinsu — DISMISSED** [review]
I will let @qqfish review this as he was working on this stuff earlier.

**Comment by whchung** [comment]
@smit-hinsu , @qqfish ping?

**Comment by whchung** [comment]
invite @yifef to this PR as it's closely related to #24295 which was just merged

**Comment by dagamayank** [comment]
Correctly tagging per @whchung comment. /cc @yifeif :)

**Comment by whchung** [comment]
Assigning the PR to @tatianashp and @thirupalanisamy as this PR is now blocking upstreaming some TensorFlow operators for ROCm.

**Comment by chsigg** [comment]
> In subsequent PRs, ROCm-specific changes for GPU device functions and GPU kernel invocation logic / macro would be introduced.

Wen-Heng, would you mind pointing me to the PRs/code that is going to come after this one. It doesn't need to be complete, just give me an idea what's going to come. As it stands, renaming the files in this PR looks like unnecessary churn because they still provide CUDA-only functionality. I assume it makes more sense once everything is merged, and I would like to get idea what that will look like.

**Comment by whchung** [comment]
@chsigg thanks. I’m working on a PR to demonstrate this. It’s taking a little more that I wanted as recently TF tip has migrated from <<<>>> to CudaLaunchKernel so our downstream fork needs to be changed a bit. I’ll manage to get a PR submitted today. In fact I think I might just amend this PR.

**Comment by whchung** [comment]
@chsigg I've revised this PR to put (many) more meats in. Macros / functions names were changed to be more vendor-neutral, and some ROCm-specific device function implementation were added.

I haven't changed kernel launching logic `CudaLaunchKernel` in this PR yet. I'd like to submit it as another PR, as this PR has already become pretty big.

**Comment by whchung** [comment]
@chsigg I've fixed all the issues from CI tests. Some failures in `Linux GPU` test target don't seem to be related to this PR. Would you mind take another look at this PR? Thanks!

**Comment by whchung** [comment]
@chsigg ping?

**Comment by chsigg** [comment]
I'm still not sold that renaming these files and function names provides real value. For example 
> `data += GpuShuffleXorSync(kCudaWarpAll, data, delta);`
is not HIP-ready, so it might be better to keep the original CUDA name.

Would you be ok with getting the _functionality_ in first, i.e. the #ifdef blocks to compute launch configs? The fact that the file and function names contain CUDA seems a cosmetic detail that we can overlook for now.

**Comment by whchung** [comment]
@chsigg all the implementation details to get ROCm path functional have already been added inside this PR:

- use device intrinsics on AMDGPU to compute shuffles
- introduce GPU_LAUNCH_KERNEL to cope with differences in kernel launch syntax between CUDA and ROCm
- use #if blocks to distinguish differences between ROCm and CUDA runtime API calls, especially mark those places yet supported by HIP runtime APIs.

- and finally terminology changes (Cuda -> GPU)

I’m wondering could you do another round of review, especially pay attention to those files in the bottommost of this PR where most of these functionality changes take place?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (chsigg) [inline tensorflow/core/util/gpu_kernel_helper.h]: Is this still needed?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (chsigg) [inline tensorflow/core/util/gpu_kernel_helper.h]: Are these still needed?

File: tensorflow/core/util/gpu_launch_config.h
Comment (chsigg) [inline tensorflow/core/util/gpu_launch_config.h]: Could you add a comment what the '1024' is about?

File: tensorflow/core/util/gpu_launch_config.h
Comment (chsigg) [inline tensorflow/core/util/gpu_launch_config.h]: Is the one block per SM limit intentional?

File: tensorflow/core/util/gpu_launch_config.h
Comment (chsigg) [inline tensorflow/core/util/gpu_launch_config.h]: 'ROCM TODO'?

**Review by chsigg — COMMENTED** [review]
I'm not familiar with how warps work on AMD. Can you run with warp size of 32 (@ half the occupancy), or do we _need_ to handle warp size 64 in the kernel implementations?

File: tensorflow/core/util/gpu_device_functions.h
Comment (whchung) [inline tensorflow/core/util/gpu_device_functions.h]: @chsigg Please refer here to see ROCm-specific logic for `GpuShuffleXorSync` for various data types.

**Comment by whchung** [comment]
@chsigg I should also mention that the purpose for this PR to is lay infrastructure for subsequent PRs for other TF operators which invokes custom HIP GPU kernels. Therefore the focus of this PR is not on the operators, but on the utility functions.

Take `GpuShuffleXorSync` as an example, you can find ROCm-specific logic in `gpu_device_functions.h`, but you won't (yet) find ROCm-specific logic in TF operators which depend on this utility function. I took this route to avoid breaking upstream implementation that much.

Shall this PR be merged, we'll then file additional PRs to enable those operators supported on ROCm incrementally like other PRs submitted and merged recently.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: This is one of the most important feature in this PR. Quite a few TF operators which depends on CUDA-specific `<<< >>>` kernel launch syntax would be changed to this macro.

Granted, I see several operators now moving toward using `CudaLaunchKernel` but still some of them employ `<<< >>>` therefore this macro is needed.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: These are needed so we have minimal impacts to the implementation of operators. Once set of `gpuXXX` macro would be built on both CUDA and ROCm.

File: tensorflow/core/util/gpu_launch_config.h
Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: Will do.

File: tensorflow/core/util/gpu_launch_config.h
Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: Will do.

File: tensorflow/core/util/gpu_launch_config.h
Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: I'll need to check with HIP runtime team on this heuristic. It was given by them before we really have `hipOccupancyMaxPotentialBlockSize` implemented in HIP runtime.

**Comment by whchung** [comment]
@chsigg On existing AMD hardware warp (wavefront in AMD terminology) is 64.

**Comment by chsigg** [comment]
> @chsigg On existing AMD hardware warp (wavefront in AMD terminology) is 64.

I'm aware of that. My question was:
> Can you run with warp size of 32 (@ half the occupancy), or do we need to handle warp size 64 in the kernel implementations?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (chsigg) [reply tensorflow/core/util/gpu_kernel_helper.h]: Ah, I see. Sounds good then. I will work towards using CudaLaunchKernel everywhere, but this is fine for now.

**Comment by whchung** [comment]
@chsigg We'll need to handle warp size of 64 in the kernel implementations. Therefore for some operators we do have to change the implementation of some kernels and they will be reflected in subsequent PRs.

File: tensorflow/core/util/gpu_launch_config.h
Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: Turns out the logic is just a clone of `GetGpuLaunchConfig(int, const Eigen::GpuDevice&)` above which assumes the kernel be simple and memory-bounded.

Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: Added comment to reflect this.

File: tensorflow/core/util/gpu_launch_config.h
Comment (whchung) [reply tensorflow/core/util/gpu_launch_config.h]: Turns out the logic is just a clone of `GetGpuLaunchConfig(int, const Eigen::GpuDevice&)`. Added comments to address that.

**Comment by whchung** [comment]
@chsigg I've addressed your review comments.

Also a ticket was raised on HIP repo some time ago to request the feature to properly estimate required launch config for HIP kernels: https://github.com/ROCm-Developer-Tools/HIP/issues/924 .

**Comment by whchung** [comment]
@chsigg rebased the PR to resolve conflicts.

Also it seems now all `<<< >>>` has been removed. I'll change `GPU_LAUNCH_KERNEL` macro to adopt `CudaLaunchKernel`.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (jlebar) [inline tensorflow/core/util/gpu_kernel_helper.h]: Why do we have both `gpuSuccess` and `GPUSUCCESS`?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (jlebar) [inline tensorflow/core/util/gpu_kernel_helper.h]: Please add some whitespace to make this easy to read; the `elif` gets lost.

**Review by jlebar — COMMENTED** [review]
Christian asked me to do a once-over on this.  LGTM with some changes to the macros to meet Google's style guide.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: after reviewing the patch, the macros aren't really necessary. I'll remove them.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: I'll remove unused `GPUSUCCESS`.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: will do.

**Review by jlebar — DISMISSED** [review]

**Comment by chsigg** [comment]
I've been trying to merge this PR, but the file and symbol renames are tricky, because we have a lot of internal references to them as well. It will be easier if I do the rename using internal tools. Then this PR will essentially add the macros and ROCm specific bits. Does that sound acceptable to you?

**Comment by whchung** [comment]
@chsigg sure thing

**Comment by whchung** [comment]
@chsigg Just want to understand where we are getting this PR merged?

**Comment by chsigg** [comment]
I'm in the process of renaming the files still. This is not trivial because we have many dependencies on these internally, some of which I don't have access to, so I need to ask others to make the change.

To unblock you, I've submitted a change so that both set of files exist, one forwarding to the other. I've not started renaming any symbols yet, but we will be in a similar situation that this is best done from our side. My recommendation is to update this PR to use the old symbols but #include the new headers, and trim it down to the functional changes that you would like to add.

**Comment by whchung** [comment]
> I'm in the process of renaming the files still. This is not trivial because we have many dependencies on these internally, some of which I don't have access to, so I need to ask others to make the change.
> 
> To unblock you, I've submitted a change so that both set of files exist, one forwarding to the other. I've not started renaming any symbols yet, but we will be in a similar situation that this is best done from our side. My recommendation is to update this PR to use the old symbols but #include the new headers, and trim it down to the functional changes that you would like to add.

got it. let me revise the PR and follow your suggestion.

/cc @jerryyin for awareness

**Comment by jerryyin** [comment]
@whchung Thanks, next weeks merging should be fun ...

**Comment by whchung** [comment]
@chsigg revised this PR so preserve existing symbols, while introducing ROCm-specific logic.

**Comment by whchung** [comment]
@chsigg a gentle ping. could you advise steps remaining to get the PR merged

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: This doesn't look correct: the static_cast will do a conversion. You probably wanted to use reinterpret_cast, which is technically not legal either.

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: Shouldn't this comment be about atomicMin/Max rather than the ordinary min/max?

File: tensorflow/core/util/gpu_device_functions.h
Comment (jlebar) [reply tensorflow/core/util/gpu_device_functions.h]: memcpy is the only legal way of doing type punning in C++.

**Comment by chsigg** [comment]
The PR description looks outdated. Would you mind to update it?

**Comment by whchung** [comment]
Thanks for the feedback. Let me try revise the PR according to your comments.

**Pull request merged**
